src/N98/Magento/Command/Github/PullRequestCommand.php

```diff
@@ -0,0 +1,140 @@
+            ->addOption('json', null, InputOption::VALUE_NONE, 'Show pull request data as json')
+            ->setDescription('Download patch from github merge request <comment>(experimental)</comment>');
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->diffContent = '';
+        $this->repository = $input->getOption('repository');
+        if ($input->getOption('mage-os')) {
+            $this->repository = 'mage-os/mageos-magento2';
+        }
+
+        $pullRequestDataResponse = $this->getPullRequestInfoByApi($input);
+
+        if ($input->getOption('json')) {
+            $output->writeln($pullRequestDataResponse->body);
+
+            return Command::SUCCESS;
+        }
+
+        $prData = $pullRequestDataResponse->decode_body(true);
+
+        if (!isset($prData['id'])) {
+            $output->writeln('<error>Could not fetch pull request data</error>');
+
+            return Command::FAILURE;
+        }
+
+        $table = PullRequestInfoTable::create($output, $prData);
+
+        /**
+         * Show only diff
+         */
+        if ($input->getOption('diff')) {
+            $output->write($this->fetchDiffContent($prData['diff_url']));
+
+            return Command::SUCCESS;
+        }
+
+        // Show infos as fallback
+        $table->render();
+
+        if ($input->getOption('patch')) {
+            $this->patchFile($prData, $output);
+        }
+
+        if (!$input->getOption('patch') && !$input->getOption('diff')) {
+            $output->writeln('Use <comment>--patch</comment> to download the patch as ready to apply patch file');
+            $output->writeln('Use <comment>--diff</comment> to see the raw diff');
+        }
+
+        return Command::SUCCESS;
+    }
+
+    /**
+     * @param string $diffUrl
+     * @return string
+     */
+    protected function fetchDiffContent($diffUrl): string
+    {
+        if ($this->diffContent === '') {
+            $response = Requests::get($diffUrl);
+            $this->diffContent = $response->body;
+        }
+
+        return $this->diffContent;
+    }
+
+    /**
+     * @param array $prData
+     * @param OutputInterface $output
+     * @return void
+     */
+    protected function patchFile(array $prData, OutputInterface $output): void
+    {
+        $patchFileContent = PatchFileContentCreator::create(
+            $this->fetchDiffContent($prData['diff_url'])
+        );
+
+        $filename = sprintf(
+            'PR-%d-%s.patch',
+            $prData['number'],
+            str_replace('/', '-', $prData['base']['repo']['full_name'])
+        );
+
+        chdir(OperatingSystem::getCwd());
+        if (file_put_contents($filename, $patchFileContent) === false) {
+            throw new \RuntimeException('Could not write patch file');
+        }
+
+        $output->writeln(sprintf('<info>Patch file created:</info> <comment>%s</comment>', $filename));
+    }
```

Comment:
Does this method handle generating multiple patch files, for each module separately? It's a bit hard to check by just reading the code, but I think it doesn't?

For example: https://github.com/magento/magento2/pull/34360
Should (in my opinion, but some people disagree) be splitted in 2 patches, for:
- `magento/magento2-base`
- `magento/module-quote-graph-ql`

Reply:
Right. PR is not ready to merge and some stuff is currently not handled. I guess we have also to manage the language packs here.

The magento/framework-message-queue package was initially part of the Enterprise Edition. Maybe that's why it's not part of the framework.

Is splitting in different patch files a good idea?

Reply:
Splitting is in my eyes necessary because you need to specify to which module a patch should apply, and if you specify from one module that some patch should apply to another module it will not work if you change the paths. 

Unless you don't change the paths and then specify a very high patch level (I think [it's 4](https://github.com/cweagans/composer-patches/pull/136) for Magento?).
Also found this super old discussion from years ago about this: https://github.com/cweagans/composer-patches/issues/76#issuecomment-394078040

I'd say, give it a try with a complex PR like I specified above and see how well it works.

Reply:
If someone uses ": "cweagans/composer-patches" then... we e.g. do not use it in Magento projects. The Quality Patches are also not splitted.

Reply:
My current plan is not to split. Splitting sounds like a lot of possible issues.

Reply:
At netz98 we have an own patch repository with a apply-patch.sh script (in the patch repo). This is then called via post-install-script/post-update-script in the `composer.json`.
Super simple solution.
Similar to the Quality Patches of Adobe.

For n98-magerun2 we should find a useful way for everyone. We use `cweagans/composer-patches` in other PHP projects, too. For Magento projects I do not like this splitting. IMHO it's a over-complication.
This does not mean that we should not support it (if it's possible).

I have not used the `vaimo/composer-patches` package. On the first look it looks for me that a standard (not splitted patch) can be used.